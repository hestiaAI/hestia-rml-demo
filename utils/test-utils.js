import DBMS from '~/utils/sql'
import FileManager from '~/utils/file-manager'

export class DatabaseTester {
  database

  async init (
    { options: { databaseConfig, preprocessors, files, keepOnlyFiles } },
    mockedFiles
  ) {
    if (!databaseConfig) {
      throw new Error('databaseConfig is not defined')
    }
    const fileManager = new FileManager(
      preprocessors,
      null,
      files,
      keepOnlyFiles
    )
    await fileManager.init(mockedFiles)
    const database = await DBMS.createDB(databaseConfig)
    const records = await DBMS.generateRecords(fileManager, databaseConfig)
    DBMS.insertRecords(database, records)

    this.database = database
  }

  select (sql) {
    return this.database.select(sql)
  }

  close () {
    this.database.close()
  }
}

export function arrayEqualNoOrder (array1, array2) {
  array2.forEach(x => expect(array1).toContainEqual(x))
  array1.forEach(x => expect(array2).toContainEqual(x))
  expect(array1.length).toBe(array2.length)
}

export function getSqlFromBlock ({ options: { viewBlocks } }, id) {
  const { sql = '' } = viewBlocks.find(b => b.id === id)
  return sql
}
