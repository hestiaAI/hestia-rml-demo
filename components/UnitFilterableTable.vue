<template>
  <div>
    <div v-if="isValid">
      <h2 v-if="$store.state.power" class="my-3">Query Results</h2>
      <v-alert v-if="error" type="error">{{ message }}</v-alert>
      <the-table-filter
        :display-filters="displayFilters"
        :headers="headers"
        @update="onFilterUpdate"
      />
      <v-data-table
        v-bind="{ headers: tableHeaders, items, search }"
        :hide-default-footer="disabled"
        multi-sort
        @current-items="onItemsUpdate"
      >
        <template #item.url="{ value }">
          <a target="_blank" rel="noreferrer noopener" :href="value"> Link </a>
        </template>
      </v-data-table>
      <base-data-download-button
        v-bind="{ progress, error, disabled, extension, data: csvData, status }"
        ref="downloadButton"
        text="Download"
        @click.native="onDownload"
      />
    </div>
    <i v-else>data in this format cannot be displayed in a table</i>
  </div>
</template>

<script>
import { writeToString } from '@fast-csv/format'
import { processError } from '@/utils/utils'
import BaseDataDownloadButton from '~/components/BaseDataDownloadButton'
import TheTableFilter from '~/components/TheTableFilter'

function isDataValid(data) {
  return !!data.items && !!(data.headers?.length > 0)
}

export default {
  name: 'UnitFilterableTable',
  components: { BaseDataDownloadButton, TheTableFilter },
  props: {
    displayFilters: {
      type: Boolean,
      default: () => false
    },
    data: {
      default: undefined,
      validator: isDataValid
    }
  },
  data() {
    return {
      status: false,
      error: false,
      progress: false,
      csvData: '',
      message: '',
      extension: 'csv',
      search: '',
      tableHeaders: this.headers
    }
  },
  computed: {
    isValid() {
      return isDataValid(this.data)
    },
    headers() {
      const rawHeaders = this.data?.headers || []
      if (typeof rawHeaders[0] === 'string') {
        // allow headers to be an array of strings
        return rawHeaders.map(h => ({ text: h, value: h }))
      }
      return rawHeaders
    },
    items() {
      return this.data?.items || []
    },
    disabled() {
      return !this.headers.length
    }
  },
  methods: {
    async onDownload(event) {
      // only if the event is trusted (i.e. triggered by user interaction)
      if (event.isTrusted) {
        // prevent default anchor behavior,
        // i.e. prevent triggering the download
        // since we want to generate the file contents first
        event.preventDefault()
        this.progress = true
        this.status = false
        this.error = false
        try {
          const headers = this.headers.map(h => h.text)
          // update the data
          this.csvData = await writeToString(this.items, { headers })
          // wait until DOM is updated, i.e. the href attribute (see BaseDownloadButton.vue)
          await this.$nextTick()
          // click the anchor manually -> event.isTrusted === false
          this.$refs.downloadButton.$el.click()
        } catch (error) {
          console.error(error)
          this.error = true
          this.message = processError(error)
        } finally {
          this.progress = false
          this.status = true
        }
      }
    },
    onFilterUpdate(selectedHeaders, searchValue) {
      this.tableHeaders = this.headers.map(h => ({
        ...h,
        filterable: selectedHeaders.includes(h.value)
      }))
      this.search = searchValue
    },
    onItemsUpdate(currentItems) {
      this.$emit('current-items', currentItems)
    }
  }
}
</script>
