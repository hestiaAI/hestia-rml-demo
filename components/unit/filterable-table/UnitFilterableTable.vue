<template>
  <div>
    <DataValidator :data="data" :allow-missing-columns="true">
      <VAlert v-if="error" type="error">{{ message }}</VAlert>
      <UnitFilterableTableFilter
        :display-filters="displayFilters"
        :headers="headers"
        @update="onFilterUpdate"
      />
      <VDataTable
        v-bind="{ headers: tableHeaders, items, search }"
        ref="tableRef"
        multi-sort
        fixed-header
        height="500"
        :footer-props="{ itemsPerPageOptions: [5, 10, 15, 500, 1000] }"
        data-testid="data-table"
        @current-items="onItemsUpdate"
      >
        <template #item.url="{ value }">
          <a target="_blank" rel="noreferrer noopener" :href="value"> Link </a>
        </template>
        <template
          v-for="header in headers.filter(
            h =>
              h.value.toLowerCase().includes('time') ||
              h.value.toLowerCase().includes('date')
          )"
          #[`item.${header.value}`]="{ value }"
        >
          {{ new Date(value).toLocaleString() }}
        </template>
      </VDataTable>
      <BaseButton
        v-bind="{ error, progress, status }"
        text="Export"
        icon="mdiExport"
        @click="exportCSV"
      />
      <BaseButtonDownloadData
        v-bind="{ disabled: !csvString, extension, data: csvString }"
      />
      <BaseButtonShare v-bind="{ disabled: !files, files }" file-share />
    </DataValidator>
  </div>
</template>

<script>
import { writeToString } from '@fast-csv/format'
import { processError } from '@/utils/utils'

export default {
  name: 'UnitFilterableTable',
  props: {
    displayFilters: {
      type: Boolean,
      default: () => false
    },
    data: {
      type: Object,
      required: true
    }
  },
  data() {
    return {
      status: false,
      error: false,
      progress: false,
      csvString: '',
      message: '',
      extension: 'csv',
      search: '',
      tableHeaders: [],
      files: null
    }
  },
  computed: {
    headers() {
      const rawHeaders = this.data?.headers || []
      if (typeof rawHeaders[0] === 'string') {
        // allow headers to be an array of strings
        return rawHeaders.map(h => ({ text: h, value: h }))
      }
      return rawHeaders
    },
    items() {
      return this.data?.items || []
    }
  },
  watch: {
    data: {
      immediate: true,
      handler(data) {
        this.tableHeaders = this.headers
      }
    }
  },
  methods: {
    async exportCSV() {
      this.progress = true
      this.status = false
      this.error = false
      try {
        const headers = this.headers.map(h => h.text)
        const filteredItems = this.$refs.tableRef.$children[0].filteredItems
        // Change the items keys to match the headers
        const itemsWithHeader = filteredItems.map(i =>
          this.headers.reduce((o, h) => ({ ...o, [h.text]: i[h.value] }), {})
        )
        // update the data
        const csv = await writeToString(itemsWithHeader, { headers })
        this.csvString = csv
        this.files = [new File([csv], 'results.csv', { type: 'text/csv' })]
      } catch (error) {
        console.error(error)
        this.error = true
        this.message = processError(error)
      } finally {
        this.progress = false
        this.status = true
      }
    },
    onFilterUpdate(selectedHeaders, searchValue) {
      this.tableHeaders = this.headers.map(h => ({
        ...h,
        filterable: selectedHeaders.includes(h.value)
      }))
      this.search = searchValue
    },
    onItemsUpdate() {
      // wait until the DOM has completely updated
      this.$nextTick(() => {
        // emit the current filtered items
        this.$emit(
          'current-items',
          // workaround to get the filtered items https://github.com/vuetifyjs/vuetify/issues/8731#issuecomment-617399086
          this.$refs.tableRef.$children[0].filteredItems
        )
      })
    }
  }
}
</script>
