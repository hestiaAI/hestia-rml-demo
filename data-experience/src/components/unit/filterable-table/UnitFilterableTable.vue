<template>
  <div>
    <DataValidator
      :data="{ items, headers }"
      allow-missing-columns
    >
      <BaseAlert v-if="error" type="error">
        {{ message }}
      </BaseAlert>
      <BaseSearchBar v-model="search" />
      <VDataTable
        v-bind="{ headers: data.headers, search }"
        ref="tableRef"
        :items="filteredItems"
        multi-sort
        :items-per-page.sync="itemsPerPage"
        fixed-header
        :height="height"
        :footer-props="{ itemsPerPageOptions: [5, 10, 15, 500, 1000] }"
        data-testid="data-table"
        @current-items="onItemsUpdate"
      >
        <template v-for="header in data.headers" #[`header.${header.value}`]>
          {{ header.text }}
          <UnitFilter
            :key="header.value"
            v-bind="{ header, values: columnValues(header) }"
            @filter-change="filterChange(header.value, $event)"
          />
        </template>
        <template #item.url="{ value }">
          <ExternalLink v-t="'Link'" :href="value" />
        </template>
        <template
          v-for="header in data.headers.filter(h => h.value !== 'url')"
          #[`item.${header.value}`]="slotProps"
        >
          {{ formatItemAsString(slotProps) }}
        </template>
      </VDataTable>
      <BaseButton
        v-bind="{ error, progress, status }"
        text="Export"
        icon="mdiExport"
        @click="exportCSV"
      />
      <BaseButtonDownloadData
        v-bind="{ disabled: !csvString, extension, data: csvString }"
      />
      <BaseButtonShare v-bind="{ disabled: !files, files }" file-share />
    </DataValidator>
  </div>
</template>

<script>
import * as Papa from 'papaparse'
import { processError } from '@/utils/utils'
import { formatObject, formatArray } from '@/utils/json'
import { TypeChecker } from '@/utils/type-check'
import kViewBlockMixin from '@/mixins/k-view-block'
import BaseButton from '@/components/base/button/BaseButton.vue'
import BaseButtonDownloadData from '@/components/base/button/download/BaseButtonDownloadData.vue'
import BaseButtonShare from '@/components/base/button/BaseButtonShare.vue'
import UnitFilter from './UnitFilter.vue'
import ExternalLink from '@/components/ExternalLink.vue'
import BaseSearchBar from '@/components/base/BaseSearchBar.vue'
import BaseAlert from '@/components/base/BaseAlert.vue'
import DataValidator from '@/components/DataValidator.vue'

const height5 = 290
const height10 = 530
const defaultItemsPerPage10 = window.innerHeight - 250 > 530

export default {
  name: 'UnitFilterableTable',
  mixins: [kViewBlockMixin],
  props: {
    headers: {
      type: Array,
      default: () => []
    },
    items: {
      type: Array,
      default: () => []
    },
    id: {
      type: String,
      required: true
    }
  },
  data() {
    return {
      advancedSearch: false,
      status: false,
      error: false,
      progress: false,
      csvString: '',
      message: '',
      extension: 'csv',
      search: '',
      files: null,
      filters: {},
      filteredItems: [],
      itemsPerPage: defaultItemsPerPage10 ? 10 : 5,
      height: defaultItemsPerPage10 ? height10 : height5
    }
  },
  computed: {
    data() {
      let tempHeaders = this.headers
      if (typeof this.headers[0] === 'string') {
        // allow headers to be an array of strings
        tempHeaders = this.headers.map(h => ({
          text: h,
          value: h
        }))
      }
      // Type detection
      const typeChecker = new TypeChecker(this)
      const { headers, items } = typeChecker.detectTypes(tempHeaders, this.items)
      // Add column style and options
      return {
        headers: headers.map(({ text, ...rest }) => {
          return {
            ...rest,
            text: this.$tev(this.kViewBlock(text, 'headers'), text),
            align: 'left',
            sortable: true
          }
        }),
        items
      }
    }
  },
  watch: {
    itemsPerPage: {
      immediate: true,
      handler(value) {
        if (value === 5) {
          this.height = height5
        } else if (defaultItemsPerPage10) {
          this.height = height10
        }
      }
    },
    data: {
      immediate: true,
      handler() {
        this.applyFilters()
      }
    }
  },
  methods: {
    filterChange(header, filter) {
      this.filters[header] = filter
      this.applyFilters()
    },
    applyFilters() {
      this.filteredItems = this.data.items.filter((d) => {
        return Object.entries(this.filters).every(([header, filter]) => {
          return filter === null || filter(d[header])
        })
      })
    },
    formatItemAsString(itemProps) {
      const { header, value } = itemProps
      // eslint-disable-next-line no-prototype-builtins
      if (header.hasOwnProperty('formatter')) {
        // call the formatter function with 'this' given as the Vue instance
        return header.formatter.call(this, value)
      }
      if (Array.isArray(value)) {
        return formatArray(value)
      }
      if (typeof value === 'object') {
        return formatObject(value)
      }
      return value
    },
    async exportCSV() {
      this.progress = true
      this.status = false
      this.error = false
      try {
        const { filteredItems } = this.$refs.tableRef.$children[0]
        // Change the items keys to match the headers
        const itemsWithHeader = filteredItems.map(i => this.data.headers.reduce((o, h) => ({ ...o, [h.text]: i[h.value] }), {}))
        // update the data
        const csv = await Papa.unparse(itemsWithHeader)
        this.csvString = csv
        this.files = [new File([csv], 'results.csv', { type: 'text/csv' })]
      } catch (error) {
        console.error(error)
        this.error = true
        this.message = processError(error)
      } finally {
        this.progress = false
        this.status = true
      }
    },
    columnValues(header) {
      return this.data.items.map(d => this.formatItemAsString({ header, value: d[header.value] }))
    },
    onItemsUpdate() {
      // wait until the DOM has completely updated
      this.$nextTick(() => {
        // workaround to get the filtered items https://github.com/vuetifyjs/vuetify/issues/8731#issuecomment-617399086
        const { filteredItems } = this.$refs.tableRef.$children[0]
        // emit the current filtered items
        this.$emit('current-items', filteredItems)
        /*
                this.$store.commit('setResult', {
                  experience: this.id,
                  result: {
                    headers: this.headers,
                    items: filteredItems
                  }
                })
                */
      })
    }
  },
  components: { BaseButton, BaseButtonDownloadData, BaseButtonShare, UnitFilter, ExternalLink, BaseSearchBar, BaseAlert, DataValidator }
}
</script>
<style>
.v-data-table-header th {
  white-space: nowrap;
}
</style>
