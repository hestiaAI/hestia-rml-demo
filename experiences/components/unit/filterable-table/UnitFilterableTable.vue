<template>
  <div>
    <DataValidator
      :data="{ items, headers }"
      allow-missing-columns
    >
      <BaseAlert v-if="error" type="error">
        {{ message }}
      </BaseAlert>
      <BaseSearchBar v-model="search" />
      <VDataTable
        v-bind="{ headers: data.headers, search }"
        ref="tableRef"
        :items="filteredItems"
        multi-sort
        :items-per-page.sync="itemsPerPage"
        fixed-header
        :height="height"
        :footer-props="{ itemsPerPageOptions: [5, 10, 15, 500, 1000] }"
        data-testid="data-table"
        @current-items="onItemsUpdate"
      >
        <template v-for="header in data.headers" #[`header.${header.value}`]>
          {{ header.text }}
          <UnitFilter
            :key="header.value"
            v-bind="{ header, values: columnValues(header) }"
            @filter-change="filterChange(header.value, $event)"
          />
        </template>
        <template #item.url="{ value }">
          <ExternalLink :href="value">
            Link
          </ExternalLink>
        </template>
        <template
          v-for="header in data.headers.filter(h => h.value !== 'url')"
          #[`item.${header.value}`]="slotProps"
        >
          {{ formatItemAsString(slotProps) }}
        </template>
      </VDataTable>
      <BaseButton
        v-bind="{ error, progress, status }"
        text="Export"
        icon="mdiExport"
        @click="exportCSV"
      />
      <BaseButtonDownloadData
        v-bind="{ disabled: !csvString, extension, data: csvString }"
      />
      <BaseButtonShare v-bind="{ disabled: !files, files }" file-share />
    </DataValidator>
  </div>
</template>

<script>
import * as Papa from 'papaparse'
import { processError } from '@/utils/utils'
import { formatObject, formatArray } from '@/utils/json'
import { detectTypes } from '@/utils/type-check'

const height5 = 290
const height10 = 530
const defaultItemsPerPage10 = window.innerHeight - 250 > 530

export default {
  name: 'UnitFilterableTable',
  props: {
    headers: {
      type: Array,
      default: () => []
    },
    items: {
      type: Array,
      default: () => []
    }
  },
  data() {
    return {
      advancedSearch: false,
      status: false,
      error: false,
      progress: false,
      csvString: '',
      message: '',
      extension: 'csv',
      search: '',
      files: null,
      filters: {},
      filteredItems: [],
      itemsPerPage: defaultItemsPerPage10 ? 10 : 5,
      height: defaultItemsPerPage10 ? height10 : height5
    }
  },
  computed: {
    data() {
      let tempHeaders = this.headers
      if (typeof this.headers[0] === 'string') {
        // allow headers to be an array of strings
        tempHeaders = this.headers.map(h => ({
          text: h,
          value: h
        }))
      }
      // Type detection
      const { headers, items } = detectTypes(tempHeaders, this.items)

      // Add column style and options
      return {
        headers: headers.map(h => ({
          ...h,
          align: 'left',
          sortable: true
        })),
        items
      }
    }
  },
  watch: {
    itemsPerPage: {
      immediate: true,
      handler(value) {
        if (value === 5) {
          this.height = height5
        } else if (defaultItemsPerPage10) {
          this.height = height10
        }
      }
    },
    data: {
      immediate: true,
      handler() {
        this.applyFilters()
      }
    }
  },
  methods: {
    filterChange(header, filter) {
      this.filters[header] = filter
      this.applyFilters()
    },
    applyFilters() {
      this.filteredItems = this.data.items.filter((d) => {
        return Object.entries(this.filters).every(([header, filter]) => {
          return filter === null || filter(d[header])
        })
      })
    },
    formatItemAsString(itemProps) {
      const { header, value } = itemProps
      // eslint-disable-next-line no-prototype-builtins
      if (header.hasOwnProperty('formatter')) {
        return header.formatter(value)
      }
      if (Array.isArray(value)) {
        return formatArray(value)
      }
      if (typeof value === 'object') {
        return formatObject(value)
      }
      return value
    },
    async exportCSV() {
      this.progress = true
      this.status = false
      this.error = false
      try {
        const { filteredItems } = this.$refs.tableRef.$children[0]
        // Change the items keys to match the headers
        const itemsWithHeader = filteredItems.map(i =>
          this.data.headers.reduce(
            (o, h) => ({ ...o, [h.text]: i[h.value] }),
            {}
          )
        )
        // update the data
        const csv = await Papa.unparse(itemsWithHeader)
        this.csvString = csv
        this.files = [new File([csv], 'results.csv', { type: 'text/csv' })]
      } catch (error) {
        console.error(error)
        this.error = true
        this.message = processError(error)
      } finally {
        this.progress = false
        this.status = true
      }
    },
    columnValues(header) {
      return this.data.items.map(d =>
        this.formatItemAsString({ header, value: d[header.value] })
      )
    },
    onItemsUpdate() {
      // wait until the DOM has completely updated
      this.$nextTick(() => {
        // workaround to get the filtered items https://github.com/vuetifyjs/vuetify/issues/8731#issuecomment-617399086
        const { filteredItems } = this.$refs.tableRef.$children[0]
        const { hash } = this.$route
        // emit the current filtered items
        this.$emit('current-items', filteredItems)

        this.$store.commit('setResult', {
          experience: hash.slice(1),
          result: {
            headers: this.headers,
            items: filteredItems
          }
        })
      })
    }
  }
}
</script>
<style>
.v-data-table-header th {
  white-space: nowrap;
}
</style>
