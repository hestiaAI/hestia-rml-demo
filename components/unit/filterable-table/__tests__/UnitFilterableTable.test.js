/**
 * @jest-environment jsdom
 */
import { mount, createLocalVue } from '@vue/test-utils'
import Vuex, { Store } from 'vuex'
import UnitFilterableTable from '../UnitFilterableTable'

const data = {
  headers: ['col1', 'col2', 'col3'],
  items: [
    {
      col1: 'Hello world',
      col2: 5,
      col3: false
    }
  ]
}

// Mock vuex store
const localVue = createLocalVue()
localVue.use(Vuex)
const store = new Store()

// Mock createObjectURL because it is not implemented in jsdom
global.URL.createObjectURL = () => {}

test('data table contains passed items', () => {
  // Mount the vue component, pass some data and a mocked vuex store
  const wrapper = mount(UnitFilterableTable, {
    propsData: {
      data
    },
    store,
    localVue
  })
  // Check that the child component exists
  expect(wrapper.find('[data-testid="data-table"]').exists()).toBe(true)
})

test('data table is not rendered when data is malformed', () => {
  // Mount the vue component, pass some data and a mocked vuex store
  const wrapper = mount(UnitFilterableTable, {
    propsData: {
      data: data.items
    },
    store,
    localVue
  })
  // Check that the child component exists
  expect(wrapper.find('[data-testid="data-error"]').exists()).toBe(false)
})

test('data table html corresponds to snapshot', () => {
  // Mount the vue component, pass some data and a mocked vuex store
  const wrapper = mount(UnitFilterableTable, {
    propsData: {
      data
    },
    store,
    localVue
  })
  // Check that the html is the same as in the snapshot
  expect(wrapper.html()).toMatchSnapshot()
})
