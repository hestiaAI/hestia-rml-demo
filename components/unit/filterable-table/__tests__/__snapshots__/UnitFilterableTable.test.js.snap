// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`data table html corresponds to snapshot 1`] = `
"<div>
  <div>
    <!---->
    <!---->
    <unitfilterabletablefilter headers=\\"[object Object],[object Object],[object Object]\\"></unitfilterabletablefilter>
    <vdatatable multi-sort=\\"\\" footer-props=\\"[object Object]\\" data-testid=\\"data-table\\" items=\\"[object Object]\\" search=\\"\\"></vdatatable>
    <basebuttondownloaddata text=\\"Download\\" extension=\\"csv\\" data=\\"\\"></basebuttondownloaddata>
  </div>
</div>"
`;
